general/development/tools/behat.md

```diff
@@ -0,0 +1,721 @@
+You can view details of all of these using the `--help` flag to `admin/tool/behat/cli/init.php`
+
+##### Running Parallel tests
+
+You can use the Moodle behat runner to run all tests, including Standard runs. It is an intelligent wrapper around the standard `./vendor/bin/behat` command which specifies the configuration file, and other required features.
+
+```
+    $ php admin/tool/behat/cli/run.php
+```
+
+Many of the standard options and parameters that can be passed to `./vendor/bin/behat` can also be passed to the Moodle runner, for example:
+
+1. `--tags` Run tests which match the specified tags
+1. `--name="Scenario name"` Run a test matching the supplied scenario name
+1. `--feature="/path/to/test.feature"` Run a specific feature file.
+1. `--suite` Features for specified theme will be executed.
+The runner also includes a number of custom parameters relating to parallel runs:
+1. `--replace` Replace args string with run process number, useful for output and reruns.
+1. `--fromrun` Execute run starting from (Used for parallel runs on different vms)
+1. `--torun` Execute run till (Used for parallel runs on different vms)
+The `--replace` feature is particularly useful and can be used to replace a string in the command with the run number. This is useful when specifying output formats, and rerun files as noted below.
+
+The following example demonstrates how Behat might be initialised with three parallel runs, to run on all themes:
+
+```
+    $ php admin/tool/behat/cli/init.php --parallel=3 --add-core-features-to-theme
+```
+
+And then to run all tests matching the `@tool_myplugin` tag, against the `classic` theme:
+
+```
+    $ php admin/tool/behat/cli/run.php --tags="@tool_myplugin" --suite="classic"
+```
+
+##### Custom parameters for parallel runs
+
+You can set following custom config options for parallel runs via $CFG->behat_parallel_run. It's an array of options where 1st array is for 1st run and so on.
+
+```php
+    $CFG->behat_parallel_run = [
+        [
+           'dbtype' => 'mysqli',
+           'dblibrary' => 'native',
+           'dbhost' => 'localhost',
+           'dbname' => 'moodletest',
+           'dbuser' => 'moodle',
+           'dbpass' => 'moodle',
+           'behat_prefix' => 'mdl_',
+           'wd_host' => 'http://127.0.0.1:4444/wd/hub',
+           'behat_wwwroot' => 'http://127.0.0.1/moodle',
+           'behat_dataroot' => '/home/example/bht_moodledata'
+       ],
+       // ...
+    ],
+```
+
+To set different selenium servers for parallel runs, you can use following.
+
+**NOTE:** Running parallel (headless) runs on different selenium servers avoid random focus failures.
+
+```php
+    $CFG->behat_parallel_run = [
+        [=> 'http://127.0.0.1:4444/wd/hub']('wd_host'),
+        [=> 'http://127.0.0.1:4445/wd/hub']('wd_host'),
+        [=> 'http://127.0.0.1:4446/wd/hub']('wd_host'),
+    ];
+```
```

Comment:
```suggestion
To set different selenium servers for parallel runs, you can use the following:

```php
    $CFG->behat_parallel_run = [
        [=> 'http://127.0.0.1:4444/wd/hub']('wd_host'),
        [=> 'http://127.0.0.1:4445/wd/hub']('wd_host'),
        [=> 'http://127.0.0.1:4446/wd/hub']('wd_host'),
    ];
```

:::tip

Running parallel (headless) runs on different selenium servers avoid random focus failures.

:::
```

Reply:
The suggestion in this case (just clarifying to guarantee it's understood) is:

- Change the order, because the $CFG->behat_parallel_run settings were not placed in the right place.
- Add a tip admonition for the NOTE.